src/Tasks/PageIterator.php

```diff
@@ -0,0 +1,190 @@
+use Http\Promise\FulfilledPromise;
+use Http\Promise\Promise;
+use InvalidArgumentException;
+use JsonException;
+use Microsoft\Graph\Core\Models\PageResult;
+use Microsoft\Kiota\Abstractions\HttpMethod;
+use Microsoft\Kiota\Abstractions\NativeResponseHandler;
+use Microsoft\Kiota\Abstractions\RequestAdapter;
+use Microsoft\Kiota\Abstractions\RequestInformation;
+use Microsoft\Kiota\Abstractions\RequestOption;
+use Microsoft\Kiota\Abstractions\Serialization\Parsable;
+
+class PageIterator
+{
+    private PageResult $currentPage;
+    private RequestAdapter $requestAdapter;
+    private bool $hasNext = false;
+    private int $pauseIndex;
+    /** @var array{string, string} $constructorFunc */
+    private array $constructorCallable;
+    private array $headers = [];
+    /** @var array<RequestOption>|null  */
+    private ?array $requestOptions = [];
+
+    /**
+     * @param mixed $response
+     * @param RequestAdapter $requestAdapter
+     * @param array{string,string} $constructorCallable The method to construct a paged response object.
+     * @throws JsonException
+     */
+    public function __construct($response, RequestAdapter $requestAdapter, array $constructorCallable) {
+        $this->requestAdapter = $requestAdapter;
+        $this->constructorCallable = $constructorCallable;
+        $this->pauseIndex = 0;
+        $page = self::convertToPage($response);
+
+        if ($page !== null) {
+            $this->currentPage = $page;
+            $this->hasNext = true;
+        }
+        $this->headers = [];
+    }
+
+    /**
+     * @param array $headers
+     */
+    public function setHeaders(array $headers): void
+    {
+        $this->headers = $headers;
+    }
+
+    /**
+     * @param array $requestOptions
+     */
+    public function setRequestOptions(array $requestOptions): void
+    {
+        $this->requestOptions = $requestOptions;
+    }
+
+    /**
+     * @param int $pauseIndex
+     */
+    public function setPauseIndex(int $pauseIndex): void
+    {
+        $this->pauseIndex = $pauseIndex;
+    }
+
+    /**
+     * @param callable $callback The callback function to apply on every entity.
+     * @throws Exception
+     */
+    public function iterate(callable $callback): void {
+        while(true) {
+            $keepIterating = $this->enumerate($callback);
+
+            if (!$keepIterating) {
+                return;
+            }
+            $nextPage = $this->next();
+
+            if (empty($nextPage)) {
+                $this->hasNext = false;
+                return;
+            }
+            $this->currentPage = $nextPage;
+            $this->pauseIndex = 0;
+        }
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function next(): ?PageResult {
+        if (empty($this->currentPage->getOdataNextLink())) {
+            return null;
+        }
+
+        $response = $this->fetchNextPage();
+        try {
+            $result = $response->wait();
+        } catch (Exception $exception){
```

Comment:
I think we should bubble up the exception so that devs know why the next page request failed.
Related to comment below on throwing `RejectedPromise` with an error message where `$nextLink` is `null`. Thoughts?